### PR TITLE
fix(copilot): expose only account-available models

### LINKED ISSUE #4697: bug(copilot): model catalog includes unavailable and unconfigured models
# Title

bug(copilot): model catalog includes unavailable and unconfigured models

# Description

### What happened

After connecting a GitHub Copilot account and refreshing its model catalog, Maka can show models that the account cannot actually use.

Two paths cause unavailable models to appear:

1. GitHub historically returns model policy states including:

   ```text
   enabled | disabled | unconfigured
   ```

   Maka currently filters only `disabled`. A model with `policy.state = "unconfigured"` therefore appears selectable even though Maka has no policy-acceptance flow and cannot use it.

2. Interactive Copilot connections initially carry the provider fallback model IDs. After account discovery, generic model reconciliation preserves previously enabled IDs even when the account-scoped `/models` response omits them. Those bootstrap IDs are then added back to the displayed catalog.

The result is that the model picker can contain:

- policy-unconfigured models;
- fallback models unavailable to the account;
- previously available models that the account no longer exposes.

Expected behavior:

- A policy-free model remains available.
- A model with `policy.state = "enabled"` remains available.
- `disabled`, `unconfigured`, unknown, and malformed policy states are excluded.
- Copilot's filtered `/models` response is treated as the account's authoritative usable catalog.
- Initial fallback IDs not present in that catalog are removed.
- Later refreshes remove withdrawn models without automatically opting the user into newly introduced models.

### How to reproduce

1. Connect a GitHub Copilot account.
2. Refresh or inspect the connection's model catalog.
3. Use a `/models` response containing a mixture such as:

   ```json
   {
     "data": [
       {
         "id": "available",
         "model_picker_enabled": true,
         "policy": { "state": "enabled" }
       },
       {
         "id": "not-configured",
         "model_picker_enabled": true,
         "policy": { "state": "unconfigured" }
       }
     ]
   }
   ```

4. Observe that `not-configured` is offered by Maka.
5. If the connection was bootstrapped with fallback IDs, observe that omitted fallback IDs can also remain selectable after discovery.

A deterministic source-level reproduction exists in the GitHub Copilot provider-contract discovery fixture by adding an `unconfigured` model. Before the fix, that entry appears in the returned `ModelInfo[]`.

### Environment

- Maka commit: `91624f1f3` or earlier affected build
- Surface: Desktop / Runtime Host model discovery
- Provider: GitHub Copilot
- OS and version: [fill in]
- Node.js version, if running from source: [fill in]

### Logs, screenshots, or additional context

GitHub's Copilot client historically represented model policy as:

```ts
type ModelPolicy = {
  state: 'enabled' | 'disabled' | 'unconfigured';
  terms?: string;
};
```

A policy-free model is also valid because newer Copilot responses may omit the retired inline policy field.

Please redact access tokens and account information from any captured `/models` response.

### PR DESCRIPTION
**Pull request by hgaol**
# Title

fix(copilot): expose only account-available models

# Description

## Summary

GitHub Copilot model discovery could expose models the connected account could not use.

This fixes both sources:

1. **Provider response filtering**
   - Policy-free models remain valid.
   - A present policy must be exactly `enabled`.
   - `disabled`, `unconfigured`, unknown, and malformed policies are excluded.
   - Existing picker, tool-call, and supported-wire requirements remain unchanged.

2. **Catalog persistence**
   - Copilot's filtered `/models` response is treated as the account's authoritative usable catalog.
   - The first account-scoped discovery replaces bootstrap fallback IDs with the available account models.
   - Later refreshes remove withdrawn IDs while preserving still-available user selections.
   - Newly introduced models remain opt-in after the initial discovery.
   - If an unavailable default must be replaced, reconciliation selects an available model.

The generic reconciliation behavior for other providers is unchanged: their model lists may be partial, so omitted user-selected IDs remain preserved.

Fixes #4697

## Root cause

`toGitHubCopilotModelInfo` rejected only:

```ts
policy?.state === 'disabled'
```

That admitted `unconfigured` models even though Maka cannot accept their policy terms.

Separately, `reconcileConnectionAfterModelFetch` intentionally preserved every previously enabled model. Interactive Copilot connections start with fallback IDs, so models absent from the account response were added back into the projected catalog.

## Verification

The regressions fail without this change and pass with it:

- Provider-contract fixture with:
  - policy-free model;
  - enabled model;
  - disabled model;
  - unconfigured model;
  - hidden-picker model;
  - tool-incompatible model;
  - unsupported endpoint.
- Core authoritative-reconciliation tests:
  - initial fallback replacement;
  - withdrawn model removal;
  - still-available user choice preservation;
  - newly introduced model remains opt-in.
- Runtime Policy integration test proving unavailable bootstrap IDs are removed from persisted Copilot state.

Checks run:

- Provider conformance matrix: **126 passed**
- `@maka/core`: **782 passed**
- `@maka/storage`: **1107 tests, 0 failures**
- `@maka/runtime`: **3176 tests, 0 failures**
- Full workspace build: passed
- Full workspace typecheck: passed
- Lint: passed
- Format check: passed
- ASF source-header check: passed

The focused regressions and full typecheck/lint/format were rerun after merging the latest `main`.

## AI use

Select exactly one:

- [ ] No generative tool made a substantive contribution
- [x] Generative tooling made a substantive contribution

Tool(s) and scope: pi with an OpenAI model assisted diagnosis, primary-source comparison against GitHub's Copilot client, implementation, test creation, and verification. The human contributor remains responsible for the change.

## Checklist

- [x] Tests cover the change and fail without it
- [x] Lint, format, typecheck and the affected suites pass locally

Does this PR entail a change in behavior?

- [x] Yes — GitHub Copilot model lists now omit models that are unavailable to the connected account.
- [ ] No

### CODE REVIEW

**Review by Astro-Han — COMMENTED** [review]
Reviewed exact head `dba7211ec39a99c12481a77a497ed6f9d68b8a29`. I found no P0–P3 issues.

What the diff actually does (7 files, +139/−17, two commits):

1. `packages/runtime/src/model-fetcher.ts` — the Copilot discovery filter now admits a model only when `policy` is absent or exactly `{ state: 'enabled' }` (`isGitHubCopilotModelPolicyEnabled`). `unconfigured`, `disabled`, malformed, and unknown states are excluded, failing closed. The existing picker, tool-call, and supported-endpoint requirements are unchanged.
2. `packages/core/src/llm-connections.ts` — `reconcileConnectionAfterModelFetch` gains an `authoritative` option: the first account-scoped fetch replaces bootstrap fallback ids with the fetched catalog, and later refreshes keep only user-enabled ids still present in the catalog, so newly introduced models stay opt-in.
3. `packages/storage/src/runtime-policy/connection-catalog-document.ts` — the single production reconcile call site sets `authoritative` only for `github-copilot`, so generic providers keep the existing preserve-everything behavior.
4. `scripts/release-cli-package.mjs` — the production dependency audit retries once on transient 5xx and transport errors; real vulnerability reports still fail closed.

What I checked:

- I read the full diff plus the surrounding reconcile logic, including alias mapping (`supersededModelId`) and the `hasModelInventory` derivation (`previous.models.length > 0`; the catalog codec forbids a non-empty `models` array before discovery, so the first Copilot fetch always takes the replace-bootstrap branch).
- Red-green per layer: with each changed file reverted to the base commit, the new core reconciliation test, the provider conformance matrix, and the storage runtime-policy integration test fail; with the PR applied all three pass. Local results on this head: core 14/14, provider contract matrix 126/126, storage runtime-policy stores 69/69.
- Hosted checks on this exact head are green: `test`, four platform addon builds, immutable tarball, four Validate-installed-CLI jobs, and Qualify released State Roots. The Validate-installed-CLI Eval job shows as skipped in a non-required matrix cell.
- No database schema change; the reconcile option is call-time only and never persisted.

Two observations, neither blocking:

- When every user-enabled model has been withdrawn but the account catalog is still non-empty, the reconciler force-enables the first catalog entry and moves the default to it (packages/core/src/llm-connections.ts:507, branch at :498). That sits in tension with the adjacent comment that newly introduced models stay opt-in (:502); the counter-argument is that it keeps the connection usable instead of leaving zero enabled models. That trade-off deserves a deliberate product decision at some point.
- An account whose entire catalog is policy-gated now fetches zero models, so a refresh (and onboarding verify) fails with the generic `invalid_response` error class (packages/runtime-host/src/server/connection-effect-coordinator.ts:158, :195). Failing closed is the right direction; the error just does not tell the user it is a policy or entitlement issue.

Housekeeping: the PR body declares a substantive generative-tool contribution, but neither commit carries a `Generated-by:` trailer. With the repository's squash-merge message setting, such a trailer would be lost at merge time regardless, which is worth noting for whoever merges.

This is not an approval; the merge decision stays with the human reviewers.

> **Automated review notice:** This comment was posted by an automated review agent operated by Astro-Han. It is not an independent human review and does not replace one.

**Comment by hgaol** [comment]
> Reviewed exact head `dba7211ec39a99c12481a77a497ed6f9d68b8a29`. I found no P0–P3 issues.
> 
> What the diff actually does (7 files, +139/−17, two commits):
> 
> 1. `packages/runtime/src/model-fetcher.ts` — the Copilot discovery filter now admits a model only when `policy` is absent or exactly `{ state: 'enabled' }` (`isGitHubCopilotModelPolicyEnabled`). `unconfigured`, `disabled`, malformed, and unknown states are excluded, failing closed. The existing picker, tool-call, and supported-endpoint requirements are unchanged.
> 2. `packages/core/src/llm-connections.ts` — `reconcileConnectionAfterModelFetch` gains an `authoritative` option: the first account-scoped fetch replaces bootstrap fallback ids with the fetched catalog, and later refreshes keep only user-enabled ids still present in the catalog, so newly introduced models stay opt-in.
> 3. `packages/storage/src/runtime-policy/connection-catalog-document.ts` — the single production reconcile call site sets `authoritative` only for `github-copilot`, so generic providers keep the existing preserve-everything behavior.
> 4. `scripts/release-cli-package.mjs` — the production dependency audit retries once on transient 5xx and transport errors; real vulnerability reports still fail closed.
> 
> What I checked:
> 
> - I read the full diff plus the surrounding reconcile logic, including alias mapping (`supersededModelId`) and the `hasModelInventory` derivation (`previous.models.length > 0`; the catalog codec forbids a non-empty `models` array before discovery, so the first Copilot fetch always takes the replace-bootstrap branch).
> - Red-green per layer: with each changed file reverted to the base commit, the new core reconciliation test, the provider conformance matrix, and the storage runtime-policy integration test fail; with the PR applied all three pass. Local results on this head: core 14/14, provider contract matrix 126/126, storage runtime-policy stores 69/69.
> - Hosted checks on this exact head are green: `test`, four platform addon builds, immutable tarball, four Validate-installed-CLI jobs, and Qualify released State Roots. The Validate-installed-CLI Eval job shows as skipped in a non-required matrix cell.
> - No database schema change; the reconcile option is call-time only and never persisted.
> 
> Two observations, neither blocking:
> 
> - When every user-enabled model has been withdrawn but the account catalog is still non-empty, the reconciler force-enables the first catalog entry and moves the default to it (packages/core/src/llm-connections.ts:507, branch at :498). That sits in tension with the adjacent comment that newly introduced models stay opt-in (:502); the counter-argument is that it keeps the connection usable instead of leaving zero enabled models. That trade-off deserves a deliberate product decision at some point.
> - An account whose entire catalog is policy-gated now fetches zero models, so a refresh (and onboarding verify) fails with the generic `invalid_response` error class (packages/runtime-host/src/server/connection-effect-coordinator.ts:158, :195). Failing closed is the right direction; the error just does not tell the user it is a policy or entitlement issue.
> 
> Housekeeping: the PR body declares a substantive generative-tool contribution, but neither commit carries a `Generated-by:` trailer. With the repository's squash-merge message setting, such a trailer would be lost at merge time regardless, which is worth noting for whoever merges.
> 
> This is not an approval; the merge decision stays with the human reviewers.
> 
> > **Automated review notice:** This comment was posted by an automated review agent operated by Astro-Han. It is not an independent human review and does not replace one.
> 

Resolved and added generated-by

**Comment by Astro-Han** [comment]
Head update note: this PR moved from `dba7211ec3` to `5f0997ff53` after the review below was posted. The new head shares both the tree and the merge-base (`26f31deafa`) with the reviewed head — the pushed commit only changed the parent, not the content — so every conclusion in the earlier review applies verbatim to `5f0997ff53`.

One unrelated but concrete reason to rebase before merging: `main` has advanced ~71 commits past the merge-base, and three of the seven files this PR touches were modified in the meantime (`packages/core/src/llm-connections.ts` — most recently via #4526 — and `scripts/release-cli-package.mjs`, three commits). That exposure is identical for both heads and does not change the review conclusions, but the overlap is worth resolving in a rebase.

---

> **Automated review notice:** This comment was posted by an automated review agent operated by Astro-Han. It is not an independent human review and does not replace one.

**Comment by Astro-Han** [comment]
Head update note: the PR head advanced to `f06bf036` via a merge of `upstream/main` into this branch (the two author commits are unchanged ancestors). The effective diff against the current base still touches the same seven files, and its content lines are byte-identical to the previously reviewed diff — only hunk offsets shifted with the base — so the conclusions in the earlier review and the head-update note above apply verbatim to `f06bf036`.

---

> **Automated review notice:** This comment was posted by an automated review agent operated by Astro-Han. It is not an independent human review and does not replace one.

**Review by Astro-Han — COMMENTED** [review]
Re-reviewed at f06bf036 (a merge of `main` into the branch; the seven files are still +139/−17 against the merged base) against `main` (`a5a99a633`), clean merge.

The Copilot half stands. `/models` already carries `policy.state` and the fetcher already dropped `disabled` (`model-fetcher.ts:526`), so treating `unconfigured` the same is the producer's call; the reconcile change is limited to `github-copilot`, whose `fallbackModels` are what kept bootstrap ids reappearing after discovery; a failed or empty fetch returns `failed` before anything is written (`connection-effect-coordinator.ts:155`), so users never lose their current models. `Generated-by` trailers are present now. Two items from last time are unchanged and stay P3: forcing the first catalogue entry on when nothing is enabled (`llm-connections.ts:505-507`), and a whole catalogue blocked by policy reporting as `invalid_response` rather than as an authorization problem.

**P2 (path ②): the second commit, the `npm audit` retry in `scripts/release-cli-package.mjs:300-325`, does not do what it says and does not belong here.** When the registry is down, `npm audit --json` prints `{"error":{"code":"E503","summary":…}}`; there is no top-level `statusCode` or `message`, and `audit.error?.code` is only set when the spawn itself fails. So `transient` is never true on the case the commit targets, and the warning prints `undefined`. The repository already has the audit authority with an `unavailable` outcome and `--allow-unavailable` (`scripts/audit-shipped-dependencies.mjs`, which parses `report.error.summary`); this adds a second transience classifier on the older `--omit=dev` path. Drop the commit from this PR. If the release path needs to tolerate a registry outage, that is a change to the existing authority in its own PR.

P3:

- `connection-catalog-document.ts:473` hard-codes `previous.providerType === 'github-copilot'` at the call site, next to `modelIdAliasesForProvider`, which is the existing seam for provider facts. A `providerReportsCompleteModelCatalog(providerType)` in `model-metadata` keeps the fact in one place.
- The removal is permanent: a model an org sets to `unconfigured` for a week comes back into `models` but not into the enabled set, and a changed default is not restored. Worth saying in the body as the accepted trade, or remembering "was enabled, now absent" ids.
- #4697 is your own issue with a source-level repro; a real `/models` response from an account with `unconfigured` entries in the body would settle the semantics.

Evidence boundary: static read; no live Copilot response was inspected; the `npm audit` error shape is from npm's documented JSON output and the repository's own parser, not from a reproduced outage.

AI-assisted review: drafted with Maka; I verified the retry predicate, the audit authority's `unavailable` path and the reconcile scope myself.

**Comment by hgaol** [comment]
Thanks for the careful re-review. Addressed at `650eb136b` on top of the latest `main`.

   **P2 — npm audit retry removed.** Agreed that this PR should not introduce a second audit/transience authority. I dropped the entire retry commit; the
 branch now has no diff in `scripts/release-cli-package.mjs` or `package.json`. Any registry-unavailable behavior belongs in `audit-shipped-dependencies.mjs`
 and a separate PR.

   **P3 — complete-catalog fact moved to the existing provider seam.** Added `providerReportsCompleteModelCatalog(providerType)` in
 `packages/core/src/model-metadata.ts`. Storage now asks that function instead of hard-coding `previous.providerType === 'github-copilot'`.

   **P3 — no forced replacement model.** Removed the fallback that enabled the first live catalog entry when every previously enabled model had disappeared.
 The reconciled state is now an empty selection/default rather than an implicit model choice. I accepted the permanence trade and updated the PR body
 accordingly: if a withdrawn model later returns, it is available but remains opt-in.

   **P3 — wholly policy-blocked catalog is authorization failure.** A non-empty Copilot catalog whose otherwise usable entries are all `disabled` or
 `unconfigured` now produces `auth`, not an empty successful list that later becomes `invalid_response`. Interactive entitlement verification maps the same
 case to `GitHubCopilotEntitlementError`.

   The cross-provider device-authorization-loop extraction remains deliberately separate, as suggested; it would rewrite two already-shipped provider paths and
 is not needed for this fix.

   Verification on the merged tree:

   - clean full build and full typecheck
   - lint and format
   - all selected standard workspace tests under Node 24
   - Runtime Host: 1785 passed, 0 failed
   - release contracts under Node 24: 196 passed
   - CLI package + installed-tarball smoke
   - focused regressions for policy-blocked discovery, authoritative empty selection, provider completeness, entitlement mapping, and persisted bootstrap-ID
 removal

   Evidence limitation remains as noted: I still have not inspected a live account response containing `unconfigured`; the behavior is pinned from GitHub
 Copilot's published source shape and deterministic source-level fixtures. A redacted real response on #4697 would close that last evidence gap.

File: packages/storage/src/runtime-policy/connection-catalog-document.ts [hunk 1]
```diff
@@ -469,6 +472,7 @@ export class ConnectionCatalogDocumentOwner {
       result.models,
       {
         aliases: modelIdAliasesForProvider(previous.providerType),
+        authoritative: providerReportsCompleteModelCatalog(previous.providerType),
```
Comment (Astro-Han) [inline packages/storage/src/runtime-policy/connection-catalog-document.ts:475]: [P1] Release a withdrawn default when committing an authoritative catalog

Normal user path: persist a Copilot inventory with `old-model` as its only enabled model and workspace default, then complete a successful nonempty fetch containing only `replacement-model`. The authoritative reconciler now correctly returns `enabledModelIds: []` and `defaultModel: ''`, but the existing code below constructs a default target with that empty ID and rejects it as outside the selection. An exact-head Storage probe produced `RuntimePolicyStoreError: Model discovery reconciled a default outside its own selection`; the snapshot retained the withdrawn old inventory/default. This also affects successful re-login followed by best-effort catalog sync.

Please handle the empty selection at this existing persistence owner: clear this connection's default target to `null` when no enabled default survives, and only migrate it when a nonempty default exists. Do not implicitly opt the replacement model in. A regression through `completeModelFetch` should begin with an existing inventory/default and assert the new catalog commits with an empty enabled set and null default.

<details><summary>中文</summary>

正常路径：已有唯一启用且作为默认的旧模型，成功刷新只返回新模型。reconciler 得到空选择，但 storage 将空字符串构造成默认目标并拒绝整次提交，实测旧目录与旧默认都被保留。建议在现有持久化权威处将已撤回的默认目标清为 null，保留新模型需要用户主动启用的语义，并用真实 completeModelFetch 覆盖该操作链。

</details>

**Review by Astro-Han — COMMENTED** [review]
Thanks for keeping account-catalog authority at the provider seam and preserving explicit model choices. One persistence-boundary case still prevents the new behavior from committing: withdrawing the only enabled/default model. Details inline.

<details><summary>中文</summary>

账号目录权威与用户选择的分工方向正确，但唯一启用/default 模型被撤回时，新结果无法通过持久化边界，详见行内。

</details>

AI-assisted review at `650eb136b`: a fresh reviewer reproduced the failure through the real Storage API; the coordinating Codex agent checked the reconciler and persistence guard.

File: packages/storage/src/runtime-policy/connection-catalog-document.ts (hunk 1 above)
Comment (hgaol) [reply packages/storage/src/runtime-policy/connection-catalog-document.ts:475]: Addressed in 4c0e44777. The storage persistence boundary now clears the connection default target to null when authoritative reconciliation produces an empty enabled selection, and only migrates it when a non-empty default survives. Added an end-to-end completeModelFetch regression covering withdrawal of the only enabled/default Copilot model. Verified with 70 runtime-policy storage tests and 15 core connection tests.

**Comment by hgaol** [comment]
Thanks for the review. Addressed the persistence-boundary case in `4c0e44777`: when authoritative Copilot discovery withdraws the only enabled/default model, the catalog now commits an empty enabled selection and clears the default target to `null`; a non-empty default is migrated as before. Added an end-to-end `completeModelFetch` regression for this scenario. Verified with 70 runtime-policy storage tests and 15 core connection tests.

File: packages/storage/src/runtime-policy/connection-catalog-document.ts (hunk 1 above)
Comment (Astro-Han) [reply packages/storage/src/runtime-policy/connection-catalog-document.ts:475]: Thanks, the source correction in `4c0e44777` addresses this P1: an empty reconciled default now becomes `null` in the same document update, while a nonempty default still has to belong to the enabled selection. The added test uses the real begin/complete fetch and default-selection APIs rather than only the pure reconciler. I checked the delta and test path; I have not independently rerun this new test, so this is not a fresh CI/merge approval.

The broader audit does not support a large architectural rewrite. Please keep the claim about current `unconfigured` account failures separate from the proven bootstrap-catalog reintroduction path: the former still needs a current redacted response and actual request/authorization evidence, as already acknowledged in the discussion. Historical client policy types do not establish today's service behavior. No new persisted selection-history mechanism is needed for the explicitly chosen opt-in behavior.

<details><summary>中文</summary>

新 head 已在同一持久化更新中把空默认值清为 null，非空默认仍验证属于 enabled；新增测试经过真实 fetch/default API。源码和测试路径已核对，本轮未独立重跑新测试，不代表新批准。体量不需要架构重写；当前 unconfigured 账号故障仍应与已证的 bootstrap 目录回填分开表述，不能用历史类型代替当前响应/请求证据，也不用新增选择历史存储。

</details>

*AI-assisted review using OpenAI Codex/Astra; evidence checked by the coordinating agent. This is not an independent human review.*

**Review by Astro-Han — APPROVED** [review]
Thanks for the fix. Re-reviewed exact head `4c0e44777cc301fd10d83d6396e01f20e2d3e504`: the persistence owner now clears a withdrawn default atomically, while nonempty defaults must still belong to the selected models. The regression goes through the real model-fetch/default-selection APIs and reads the committed snapshot. The reported P1 is resolved; I found no remaining merge-blocking issue in this revision. Exact-head CI passed.

The previously discussed boundary remains: current `unconfigured` account behavior has not been independently sampled, and newly returned models remain opt-in. This review does not claim a live-account verification. The small documentation clarification can follow without adding another selection authority.

<details><summary>中文</summary>

最新 head 已在持久化权威处原子清理被撤回的默认模型，非空默认仍须属于选择集合。回归经过真实 fetch/default API 并读取提交结果，P1 已修，当前 CI 通过。未发现本次修订剩余阻塞问题。仍未独立采样真实账号的 unconfigured 行为，重新出现的模型保持手动启用；不因此新增选择历史权威。

</details>

AI-assisted review using OpenAI Codex/Astra. The coordinating agent checked the source delta and test path; the new test was not rerun locally. This is not an independent human review.
